File: Java/백준/Silver/9012. 괄호/괄호.java
import java.util.Scanner;
import java.util.Stack;

public class Main {

	public static void main(String[] args) {
		StringBuilder sb = new StringBuilder();
		Scanner scan = new Scanner(System.in);
		int T = Integer.parseInt(scan.nextLine());
		L : for(int i = 0; i < T; i ++) {
			char[] line = scan.nextLine().toCharArray();
			Stack<Character> stack = new Stack<>();
			for(char c : line) {
				if(c == '(')
					stack.push(c);
				else {
					if(stack.isEmpty()) {
						sb.append("NO").append('\n');
						continue L;
					}
					else stack.pop();
				}
			}
			if(!stack.isEmpty()) sb.append("NO").append('\n');
			else sb.append("YES").append('\n');
		}
		System.out.println(sb);
	}

}

File: Kotlin (JVM)/백준/Silver/9012. 괄호/괄호.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main(args: Array<String>){
    var sb: StringBuilder = StringBuilder()
    var scan: BufferedReader = BufferedReader(InputStreamReader(System.`in`))
    var stack: Stack<Char> = Stack()
    var N: Int = scan.readLine().toInt()
    loop@ for (i: Int in 0 until N){
        stack.clear()
        var data: CharArray = scan.readLine().toCharArray()
        for (c: Char in data){
            if (c == '(') stack.push(c)
            else {
                if(stack.empty()){
                    sb.append("NO\n");
                    continue@loop
                } else stack.pop()
            }
        }
        if(!stack.empty())sb.append("NO\n");
        else sb.append("YES\n");
    }
    println(sb)
}